### PR TITLE
Fix marks incorrect behavior on insertText

### PR DESCRIPTION
**Pull request by sgurenkov**
#### Is this adding or improving a _feature_ or fixing a _bug_?

_bug_

#### What's the new behavior?
before:
before

after:
after

#### How does this change work?

`insertTextByPath` should replace marks on the inserted text with the once that are provided

It were trying to add those marks that are passed to the function to the existing once,
but it should replace them instead.

Example error behavior was:
* put cursor at the end of the marked text
* toggle marks
* enter text

expected:
text is being added without toggled marks

actual:
text is added with those marks applied

#### Have you checked that...?

* [x] The new code matches the existing patterns and styles.
* [x] The tests pass with `yarn test`.
* [x] The linter passes with `yarn lint`. (Fix errors with `yarn prettier`.)
* [x] The relevant examples still work. (Run examples with `yarn watch`.)

#### Does this fix any issues or need any specific reviewers?

Fixes: #2848 
Reviewers: @adjourn, @erquhart

### CODE REVIEW

**Comment by erquhart** [comment]
This looks like it strikes a good balance! Still a breaking change, but there's really no way around that for a proper fix. I'll try it out soon. Thanks so much for putting this together!

**Comment by nimishadoongarwal** [comment]
We are using slate editor and the functionality has been broken for a while, wanted to get this fixed, is it possible to merge this PR?

**Comment by erquhart** [comment]
Needs to be tested - can you pull it in to your project locally and report back if there are any issues? That will help give some confidence to move this along.

**Comment by sgurenkov** [comment]
yeah, hopefully we can get it in soon, there is one more fix coming related to undo/redo and selection events

**Comment by adjourn** [comment]
I'm very busy right now but if I get around working on text editor again in my project (hopefully this week), I'll give it a proper look and test run. Thank you @sgurenkov for putting this together.

**Comment by arendora** [comment]
This is really awesome @sgurenkov so thanks for putting it together! We are also using slate in our project at the moment as well so looking forward to this fix.

I've tried out this change with our project, and it looks like there's still an edge case issue whereby, when you are the very end of the text (no other text/media after it) and you try toggling the markdown off, it remains on.

### Scenario
**When** you are typing at the end of the whole text in the editor
**And** the text is formatted
**And** you `toggleMark`
**And** you continue typing

### Expected
**Then** the new text is no longer formatted in that mark

### Observed
**Then** the new text is still formatted in that mark

**Comment by sgurenkov** [comment]
@arendora Hello, thank you for the feedback, I've tried to replicate your scenario, but cannot catch that issue

test

can you please provide particular steps to reproduce the issue?

**Comment by arendora** [comment]
tl;dr I never had your changes to begin with. Sorry for the false alarm, no problems with this PR.

@sgurenkov Sorry for flagging a non-existent issue 😓 looks like the problem is on my side from trying to specify my slate dependency version from your branch and repo (i.e. github.com:docusign/slate.git#fix-marks-bugs) which [does not work] 🤦‍♀ (https://github.com/yarnpkg/yarn/issues/4725) . I ran and linked your changes locally and it all looks good with our codebase.

Looking forward to this PR being merged

**Comment by sgurenkov** [comment]
@arendora no problem, thank you for verifying! 👍🏻

**Comment by Calyhre** [comment]
This works for us too. Thanks for that fix 👍

**Comment by gustafguner** [comment]
Works here too, thank you for the bug fix! 🙏

**Comment by 0ctothorp** [comment]
I just wanted to approve that this fix also works for me. Thanks!

**Comment by drivasperez** [comment]
Hi, thanks for this fix. 

In the meantime before this PR is merged, do we know which versions of the various slate packages we should revert to, to avoid this issue?

**Comment by ianstormtaylor** [comment]
Thank you for this @sgurenkov! Great quality PR.
